### ISSUE TITLE
DPO/KTO training goes NaN with the default SDPA backend on torch 2.13 (cuDNN attention); clean with `enable_cudnn_sdp(False)` or eager

### ISSUE THREAD

**Issue by qgallouedec** [huggingface/trl#7181]
### Reproduction

DPO on Qwen2.5-0.5B-Instruct, bf16, `attn_implementation="sdpa"`, 64 steps. The loss looks normal for tens of steps, then `grad_norm` turns NaN and never recovers. Same script with the cuDNN SDPA backend disabled is clean.

```python
# repro_cudnn_nan.py
#   python repro_cudnn_nan.py            -> grad_norm=nan within ~60 steps
#   python repro_cudnn_nan.py --no_cudnn -> clean
import argparse
import torch
from datasets import load_dataset
from transformers import AutoModelForCausalLM, AutoTokenizer
from trl import DPOConfig, DPOTrainer

parser = argparse.ArgumentParser()
parser.add_argument("--no_cudnn", action="store_true")
args = parser.parse_args()
if args.no_cudnn:
    torch.backends.cuda.enable_cudnn_sdp(False)
print("cudnn_sdp_enabled:", torch.backends.cuda.cudnn_sdp_enabled())

model_id = "Qwen/Qwen2.5-0.5B-Instruct"
tokenizer = AutoTokenizer.from_pretrained(model_id)
model = AutoModelForCausalLM.from_pretrained(model_id, dtype=torch.bfloat16, attn_implementation="sdpa")
dataset = load_dataset("trl-lib/ultrafeedback_binarized", split="train").shuffle(seed=0).select(range(2048))

config = DPOConfig(
    output_dir="repro_out",
    max_length=768,
    per_device_train_batch_size=4,
    gradient_accumulation_steps=8,
    learning_rate=1e-6,
    max_steps=64,
    bf16=True,
    gradient_checkpointing=True,
    logging_steps=1,
    save_strategy="no",
    report_to="none",
)
trainer = DPOTrainer(model=model, args=config, train_dataset=dataset, processing_class=tokenizer)
trainer.train()
nan_steps = [h["step"] for h in trainer.state.log_history if "grad_norm" in h and h["grad_norm"] != h["grad_norm"]]
print("steps with grad_norm=nan:", nan_steps[:5], "..." if len(nan_steps) > 5 else "", f"({len(nan_steps)}/64)")
```

### Outcome

```
$ python repro_cudnn_nan.py
cudnn_sdp_enabled: True
steps with grad_norm=nan: [58, 59, 60, 61, 62] ... (7/64)

$ python repro_cudnn_nan.py --no_cudnn
cudnn_sdp_enabled: False
steps with grad_norm=nan: []  (0/64)
```

The step at which it breaks depends on the batch: with the same setup on other 2k-example subsets I got the first NaN at step 27; on a 24k-example run at step 10-70. Once `grad_norm` is NaN, loss is logged as 0 for the rest of training. Same with `gradient_checkpointing=False`. `attn_implementation="eager"` is clean too. KTOTrainer shows the same behaviour (NaN at step 2 on a 256-example run, clean with eager).

The forward is already off before the NaN: on a KTO batch the logged `kl` at step 1 is 0.137 with cuDNN on vs 0.078 with eager, so it is not only a backward problem.

### Expected behavior

Finite gradients regardless of the SDPA backend torch picks. Since torch 2.13 enables the cuDNN attention backend by default on Hopper, users running the standard DPO/KTO examples hit this with default settings, and short runs (< ~30 steps) don't show it.

### System Info

- Platform: Linux-6.8.0-1061-aws-x86_64-with-glibc2.35
- Python version: 3.12.13
- TRL version: 1.13.0
- PyTorch version: 2.13.0 (cu130)
- accelerator(s): NVIDIA H100 80GB HBM3
- Transformers version: 5.16.1
- Accelerate version: 1.15.0
- Accelerate config: not found
- Datasets version: 5.0.1
- HF Hub version: 1.31.0
- bitsandbytes version: not installed
- DeepSpeed version: not installed
- Liger-Kernel version: not installed
- PEFT version: 0.20.0
- vLLM version: 0.28.0

### Checklist

- [x] I have checked that my issue isn't already filed (see [open issues](https://github.com/huggingface/trl/issues?q=is%3Aissue))
- [x] I have included my system information
- [x] Any code provided is minimal, complete, and reproducible ([more on MREs](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks))
- [x] Any code provided is properly formatted in code blocks, (no screenshot, [more on code blocks](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks))
- [x] Any traceback provided is complete

**Comment by qgallouedec** [comment]
Root cause found.

It is a cuDNN SDPA backward kernel bug, not a TRL or transformers bug. Reduced to a 12-line torch-only repro; the DPO NaN is this bug hit through Qwen2.5's "massive activation" attention head.

Repro:

```python
import math, torch, torch.nn.functional as F
torch.manual_seed(0)
L, D = 704, 64  # fails for every L % 128 == 64; 640 and 768 are clean
q = torch.randn(1, 1, L, D, device="cuda", dtype=torch.bfloat16) * 0.1
k = torch.randn(1, 1, L, D, device="cuda", dtype=torch.bfloat16) * 0.1
v = torch.randn(1, 1, L, D, device="cuda", dtype=torch.bfloat16)
q[..., 0], k[..., 0] = 1.0, -120 * math.sqrt(D)  # every logit ~ -120, so every row's logsumexp < -88.7
q.requires_grad_(); k.requires_grad_(); v.requires_grad_()
mask = torch.ones(L, L, dtype=torch.bool, device="cuda").tril()[None, None]  # an all-True mask fails too
out = F.scaled_dot_product_attention(q, k, v, attn_mask=mask)
out.backward(torch.randn_like(out))
print("grad_q non-finite:", (~torch.isfinite(q.grad)).sum().item())  # 45056; 0 with enable_cudnn_sdp(False) or is_causal=True
```

Mechanism (bisected from the failing DPO micro-batch, Qwen2.5-0.5B layer 8 head 3):

- Forward is fine (cuDNN err vs fp32 = mem-efficient err). Only `grad_q` goes NaN. Same q/k/v/mask through EFFICIENT or MATH: clean.
- Needs three things at once: an explicit `attn_mask` (all-True triggers too, so it is the bias code path), `seq_len % 128 == 64`, and a query row with logsumexp < -88.7 = ln(FLT_MAX). The kernel scores the padded half of the last 128-wide key tile as 0 instead of -inf, `exp(0 - lse)` overflows fp32, and inf * 0 = NaN in dQ. Above the threshold the garbage is finite and multiplied by zero, hence invisible.
- Layer 8 head 3 has logits in the +-1600 range, so most of its rows qualify; any padded DPO/KTO batch of length 64 mod 128 then NaNs. Hence batch dependent and tens of steps.
- Reproduces on torch 2.13.0 cu130/cu129 (cuDNN 9.20) and torch 2.11.0 (cuDNN 9.19). `check_prefer_cudnn_attention` puts cuDNN first when cuDNN > 9.15 on SM90/SM100, even though `torch._C._get_sdp_priority_order()` still prints it last. So it is "since cuDNN > 9.15", not "since torch 2.13".

Workarounds, verified: `torch.backends.cuda.enable_cudnn_sdp(False)`, or `TORCH_CUDNN_SDPA_DEPRIORITIZED=1`, or `attn_implementation="eager"`. Nothing to change in TRL. Will file upstream with the repro and link it here.
